### ISSUE TITLE
ResolveEngine.kt is getting big

### ISSUE THREAD

**Issue by mkaput** [intellij-rust/intellij-rust#1013]
Maybe this is only my personal feeling, but I think [ResolveEngine.kt](https://github.com/intellij-rust/intellij-rust/blob/723969c995e7991e1bc002454656f1cdb27edd2f/src/main/kotlin/org/rust/lang/core/resolve/ResolveEngine.kt) becomes too big. It has already > 500 lines, and it's growing.

As for now this file contains an object with multiple not-so-strictly coupled methods whose only similarity is their use case (and millions of millions of various utility functions). Some of them have even non consistent api, see:

```kotlin
fun resolveFieldExpr(fieldExpr: RsFieldExpr): List<RsCompositeElement>
fun resolveMethodCallExpr(call: RsMethodCallExpr): RsNamedElement?
fun resolveCallExpr(path: RustPath, element: RsPath, namespace: Namespace?): List<RsCompositeElement>
fun resolveStructExprField(structExpr: RsStructExpr, fieldName: String): List<RsNamedElement>
fun resolveUseGlob(ref: RsUseGlob): List<RsCompositeElement>
fun resolveModDecl(modDecl: RsModDeclItem): RsNamedElement?
fun resolveExternCrate(crate: RsExternCrateItem): RsNamedElement?
fun resolveLabel(label: RsLabel): RsLabelDecl?
fun resolveLifetime(lifetimeRef: RsLifetime): RsLifetimeDecl?
```

All these methods are named with pattern `resolve{PSI element name}` and so one would expect them to have also some pattern API, which they do not.

Maybe we should refactor resolve engine before it becomes total nightmare?

First of all, I think, three things have to be separated:
 - resolving by name (string) - maybe this should be converted to project service? Also current API is pretty poor (or I don't understand how to use it); how to resolve fully qualified name without knowing in which module it is?
 - resolving PSI elements per se - maybe these ones should be converted to extension methods?
 - utils

**Comment by matklad** [comment]
Yeah, ResolveEngine is too unwieldy at the moment. 

I don't necessary agree with the size concern though: a bulk of this file is a rather dense meaty code, which actually does computation (as opposed to a ton of indirection sometimes found in OO code bases), and it's nice to keep or increase this density.

But it surely would be nice to get rid of those useless methods you've mentioned, and from a ton of reference classes as well. Also, the namespace API is just wrong (namespaces should be pushed down to the resolve engine) and method related API is wrong (It'd be nice to express everything in terms of UFCS). 

>how to resolve fully qualified name without knowing in which module it is?

Hm, you absolutely do need to know a module to do anything. Nothing exists outside a module. PsiElements are implicitly bound to a module, but if you don't have a PsiElement, you should provide a module yourself. But in general I think that a better strategy would be to create a fake PSI element and run the usual resolve on it, rather than do some custom thing.

**Comment by mkaput** [comment]
> Hm, you absolutely do need to know a module to do anything. Nothing exists outside a module. PsiElements are implicitly bound to a module, but if you don't have a PsiElement, you should provide a module yourself. But in general I think that a better strategy would be to create a fake PSI element and run the usual resolve on it, rather than do some custom thing.

I currently put the thing which used it on hold and I'll investigate it later, but the case was, that the only things I got were: fully qualified name, project and global search scope

**Comment by matklad** [comment]
One more interesting though that has not occurred to me recently: current ResolveEngine is impossible to profile, because all you would see is a tone of deeply recursive internal `Sequence` methods: filtering sequence calls filtering sequence calls flattening sequence ... 

It's interesting that `ScopeProcessor` solution, where you basically push down a callback `(ScopeEntry) -> Boolean` will give you nice stack, profile an debugging experience ...

**Comment by matklad** [comment]
I've revamped out resolve engine :) 

The major change is that it is now based on scope processors, instead of sequences (instead of constructing a lazy `Sequence`, we pass a callback). This gives **much** nicer stacktraces, is actually more lazy and allows to implement stateful bits like shadowing in a more natural way. Some other improvements:

* Shadowing and multi resolve now work by design, and not by accident. No more random `.firstOrNull` calls.

* Completion and resolve now fully share code (as a bonus, we've got better completion for `self::` and `super::` without a separate completion contributror, and we've also got completion for lifetimes).

* Namespaces are pushed down during resolve, which allows to get rid of some hacks. 

* String based resolve now works by creating a fake PSI element and resolving it. 

We still have a zillion of methods and almost identical Reference classes though :(
